app/models/package_manager/pypi/json_api_project.rb

```diff
@@ -0,0 +1,109 @@
+module PackageManager
+  class Pypi
+    class JsonApiProject
+      def self.request(project_name:)
+        data = begin
+          ApiService.request_json_with_headers("https://pypi.org/pypi/#{project_name}/json")
+        rescue StandardError => e
+          Rails.logger.error("Invalid PyPI JSON API Project: #{project_name}: #{e.message}")
+
+          {}
+        end
+
+        new(data)
+      end
+
+      def self.release_data_published_at(details)
+        return nil if details == []
+
+        upload_time = details.dig(0, "upload_time")
+        raise ArgumentError, "does not contain upload_time" unless upload_time
+
+        return nil unless upload_time
+
+        begin
+          Time.parse(upload_time)
+        rescue ArgumentError => e
+          Rails.logger.error("Unable to parse PyPI JSON API Project upload_time: #{project_name} (#{upload_time}): #{e.message}")
+
+          nil
+        end
+      end
+
+      def initialize(data)
+        @data = data
+      end
+
+      def license
+        @data.dig("info", "license")
+      end
+
+      def license_classifiers
+        license_classifiers = @data.dig("info", "classifiers").select { |c| c.start_with?("License :: ") }
+        license_classifiers.map { |l| l.split(":: ").last }.join(",")
+      end
+
+      def name
+        @data.dig("info", "name")
+      end
+
+      def description
+        @data.dig("info", "summary")
+      end
+
+      def homepage
+        @data.dig("info", "home_page")
+      end
+
+      def keywords_array
+        Array.wrap(@data.dig("info", "keywords").try(:split, /[\s.,]+/))
+      end
+
+      def licenses
+        license.presence || license_classifiers
+      end
+
+      def repository_url
+        ["Source", "Source Code", "Repository", "Code"].filter_map do |field|
+          @data.dig("info", "project_urls", field)
+        end.first
+      end
+
+      def homepage_url
+        @data.dig("info", "home_page").presence ||
+          @data.dig("info", "project_urls", "Homepage")
+      end
+
+      def preferred_repository_url
+        RepositoryService.repo_fallback(
+          repository_url,
+          homepage_url
+        )
+      end
+
+      def releases
+        JsonApiProjectReleases.new(
+          @data["releases"].map do |version_number, details|
+            JsonApiProjectRelease.new(
+              version_number: version_number,
+              published_at: self.class.release_data_published_at(details)
```

Comment:
Could `release_data_published_at` be an instance method? I only see it being used here

Reply:
I would want to make it a private method and then add tests to `#releases` to ensure all the private method's cases are handled.

Reply:
Scratch that, this is going to need more refactoring to be structured correctly.